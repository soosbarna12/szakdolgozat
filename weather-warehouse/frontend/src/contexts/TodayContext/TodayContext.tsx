import React, { createContext, useEffect, useState } from "react";
import { TodayContextType, TodayLocationData, TodayProviderProps } from "./TodayContext.type";


export const TodayContext = createContext<TodayContextType>({
	location: { name: "", lat: 0, lon: 0 },
	setLocation: () => { },
});

export const TodayLocationProvider = ({ children }: TodayProviderProps) => {

	const [location, setLocation] = useState<TodayLocationData>(() => {
		const storedLocation = localStorage.getItem("todayLocation");
		const defaultLocation = { name: "", lat: 0, lon: 0 };
		try {
			return storedLocation ? JSON.parse(storedLocation) : defaultLocation;
		} catch (error) {
			return defaultLocation;
		}
	});

	useEffect(() => {
		localStorage.setItem("todayLocation", JSON.stringify(location));
	}, [location]);


	return (
		<TodayContext.Provider value={{ location, setLocation }}>
			{children}
		</TodayContext.Provider>
	);
};
