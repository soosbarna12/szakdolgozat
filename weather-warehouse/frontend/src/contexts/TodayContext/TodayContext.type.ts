import { Dispatch, ReactNode, SetStateAction } from "react";

// location name interfere
export interface TodayLocationData {
  name: string;
  country?: string;
  state?: string;
  lat: number;
  lon: number;
}

export interface TodayContextType {
  location: TodayLocationData;
  setLocation: Dispatch<SetStateAction<TodayLocationData>>;
}

export interface TodayProviderProps {
	children: ReactNode;
}
