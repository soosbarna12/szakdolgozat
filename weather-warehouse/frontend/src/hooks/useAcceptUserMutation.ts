import { useMutation } from "@tanstack/react-query";
import { useAlert } from "../utils/AlertContext";
import axios from "../utils/axiosConfig";


export function useAcceptUserMutation() {
  const { showAlert } = useAlert();

  const mutation = useMutation({
    mutationFn: async (id: number) => {
      const response = await axios.post(`/user/accept`, { id });
      return response.data;
    },
    onSuccess: () => {
      showAlert("User accepted successfully", "success");
    },
    onError: () => {
      showAlert("Failed to accept user", "error");
    },
  });

  return mutation;
}
