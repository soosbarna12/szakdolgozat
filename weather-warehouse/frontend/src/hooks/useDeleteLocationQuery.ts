import { useMutation } from "@tanstack/react-query";
import { useAlert } from "../utils/AlertContext";
import axios from "../utils/axiosConfig";

export function useDeleteLocationQuery() {
  const { showAlert } = useAlert();

  const mutation = useMutation({
    mutationFn: async (userLocationID: number) => {
      const response = await axios.delete(`/user/deleteLocation`, { params: { userLocationID } });
      return response.data;
    },
    onSuccess: () => {
      showAlert("Location deleted successfully", "success");
    },
    onError: () => {
      showAlert("Failed to delete location", "error");
    },
  });

  return mutation;
}