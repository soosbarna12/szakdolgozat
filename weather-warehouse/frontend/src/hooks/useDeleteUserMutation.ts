import { useMutation } from "@tanstack/react-query";
import { useAlert } from "../utils/AlertContext";
import axios from "../utils/axiosConfig";


export function useDeleteUserMutation() {
  const { showAlert } = useAlert();

  const mutation = useMutation({
    mutationFn: async (id: number) => {
      const response = await axios.delete(`/user/delete`, { params: { id } });
      return response.data;
    },
    onSuccess: () => {
      showAlert('User deleted successfully', 'success');
    },
    onError: () => {
      showAlert('Failed to delete user', 'error');
    },
  });

  return mutation;
}
