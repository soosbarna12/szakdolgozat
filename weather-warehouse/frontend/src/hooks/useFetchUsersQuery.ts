import { useQuery } from "@tanstack/react-query";
import { User } from "../types/userData.type";
import axios from "../utils/axiosConfig";


export function useFetchUsersQuery() {
  
  const { data, error, isLoading, refetch } = useQuery<User[]>({
    queryKey: ["users"],
    queryFn: async () => {
      const response = await axios.get(`/api//user/userData`);

      return response.data as User[];
    },
  });

  return { data, error, isLoading, refetch };
}
