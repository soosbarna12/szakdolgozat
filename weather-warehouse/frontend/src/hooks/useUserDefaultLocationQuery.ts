import { useQuery } from "@tanstack/react-query";
import { useEffect } from "react";
import { useAlert } from "../utils/AlertContext";
import axios from "../utils/axiosConfig";


export function useUserDefaultLocationQuery(lat: number | undefined, lon: number | undefined) {
  const { showAlert } = useAlert();
  
  const { data, error, isLoading } = useQuery({
    queryKey: ["userDefaultLocation", lat, lon],
    queryFn: async () => {
      const response = await axios.get(`/api/today/reverse-geocode`, {
        params: { lat, lon },
      });

      return response.data;
    },
    enabled: !!(lat && lon),
    retry: 1,
  });

  useEffect(() => {
    if (error) {
      showAlert("Failed to fetch reverse geocoded location:", "error");
    }
  }, [error]);

  return { data, error, isLoading };
}
