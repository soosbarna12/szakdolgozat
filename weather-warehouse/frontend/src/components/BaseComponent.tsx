import React, { useContext, useEffect, useState } from "react";
import { Route, Routes } from "react-router-dom";
import { ROUTES } from "../consts/routes";
import { TodayContext } from "../contexts/TodayContext/TodayContext";
import { useUserDefaultLocationQuery } from "../hooks/useUserDefaultLocationQuery";
import { TodayPage } from "../pages/TodayPage/TodayPage";


export function BaseComponent() {
  const [coords, setCoords] = useState<GeolocationCoordinates>();
  const { location, setLocation } = useContext(TodayContext);
  const { data } = useUserDefaultLocationQuery(coords?.latitude, coords?.longitude);

  useEffect(() => {
    if (data && coords) {
      setLocation({
        name: data.name,
        country: data.country,
        lat: coords.latitude,
        lon: coords.longitude,
      });
    }

  }, [data]);


  // get user's default location
  useEffect(() => {
    if (navigator.geolocation && !location?.name) {
      navigator.geolocation.getCurrentPosition(
        async (position) => {
          setCoords(position.coords);
        }
      )
    }
  }, []);


  function renderRouteElements() {
    return ROUTES.map(({ id, path, element }) => (
      <Route key={id} path={path} element={element} data-testid={`route${id}`} />
    ));
  }

  return (
    <>
      <Routes>
        <Route path="/">
          <Route index element={<TodayPage />} data-testid="routeToday" />
          {renderRouteElements()}
          <Route path="*" element={<TodayPage />} data-testid="routeNotFound" />
        </Route>
      </Routes>
    </>
  );
}
