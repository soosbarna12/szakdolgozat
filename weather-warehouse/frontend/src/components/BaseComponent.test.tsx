import '@testing-library/jest-dom';
import { waitFor } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { TodayContext } from '../contexts/TodayContext/TodayContext';
import { useUserDefaultLocationQuery } from '../hooks/useUserDefaultLocationQuery';
import { renderWithQueryClient } from '../utils/test/renderWithQueryClient';
import { BaseComponent } from './BaseComponent';

jest.mock('../hooks/useGeolocationQuery', () => ({
  useGeolocationQuery: jest.fn(() => ({
    data: [],
    error: null,
    isLoading: false,
  })),
}));

jest.mock('../hooks/useUserDefaultLocationQuery', () => ({
  useUserDefaultLocationQuery: jest.fn(() => ({
    data: null,
    error: null,
    isLoading: false,
  })),
}));

jest.mock('react-leaflet', () => ({
  MapContainer: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="dataMapContainer">{children}</div>
  )
}));

jest.mock('export-to-csv', () => ({
  download: jest.fn(),
  generateCsv: jest.fn(),
  mkConfig: jest.fn(),
}));

describe('BaseComponent', () => {
  const mockSetLocation = jest.fn();
  const mockGetCurrentPosition = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    (useUserDefaultLocationQuery as jest.Mock).mockReturnValue({ data: null });

    Object.defineProperty(global.navigator, 'geolocation', {
      value: {
        getCurrentPosition: mockGetCurrentPosition,
      },
      writable: true,
    });
  });

  const renderComponent = (location = { name: "Budapest", country: "HU", lat: 0, lon: 0 }) => {
    return renderWithQueryClient(
      <MemoryRouter>
        <TodayContext.Provider value={{ location, setLocation: mockSetLocation }}>
          <BaseComponent />
        </TodayContext.Provider>
      </MemoryRouter>
    );
  };

  it('matches BaseComponent snapshot', () => {
    const component = renderComponent();
    expect(component).toMatchSnapshot();
  });

  it('calls geolocation API when location is not set', async () => {
    mockGetCurrentPosition.mockImplementationOnce((successCallback) => {
      successCallback({ coords: { latitude: 40.7128, longitude: -74.006 } });
    });

    renderComponent({ name: null, country: null, lat: 0, lon: 0 });

    await waitFor(() => {
      expect(mockGetCurrentPosition).toHaveBeenCalled();
    });
  });
});