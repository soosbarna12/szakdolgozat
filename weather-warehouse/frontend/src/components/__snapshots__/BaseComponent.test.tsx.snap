// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseComponent matches BaseComponent snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorTransparent MuiAppBar-positionStatic css-94iixj-MuiPaper-root-MuiAppBar-root"
        style="--Paper-shadow: none;"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-t4zkma-MuiToolbar-root"
        >
          <div
            class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1tlcqt-MuiAutocomplete-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-rs1at4-MuiFormControl-root-MuiTextField-root"
              data-testid="locationSearchBox"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1e3qgpi-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1nowbqt-MuiInputAdornment-root"
                >
                  <span
                    aria-hidden="true"
                    class="notranslate"
                  >
                    ​
                  </span>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                    data-testid="SearchOutlinedIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
                    />
                  </svg>
                </div>
                <input
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-invalid="false"
                  autocapitalize="none"
                  autocomplete="off"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-qve9l1-MuiInputBase-input-MuiOutlinedInput-input"
                  id="«r0»"
                  placeholder="Budapest"
                  role="combobox"
                  spellcheck="false"
                  type="text"
                  value="Budapest, HU"
                />
                <div
                  class="MuiAutocomplete-endAdornment css-1uhhrmm-MuiAutocomplete-endAdornment"
                >
                  <button
                    aria-label="Open"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-2y136s-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                    tabindex="-1"
                    title="Open"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                      data-testid="ArrowDropDownIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                  </button>
                </div>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-w4cd9x"
                  >
                    <span
                      aria-hidden="true"
                      class="notranslate"
                    >
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
      </header>
      <div
        class="MuiBox-root css-g04nnx"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-wave css-jk665a-MuiSkeleton-root"
          data-testid="todaySkeleton"
          style="width: 350px; height: 375px;"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorTransparent MuiAppBar-positionStatic css-94iixj-MuiPaper-root-MuiAppBar-root"
      style="--Paper-shadow: none;"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-t4zkma-MuiToolbar-root"
      >
        <div
          class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1tlcqt-MuiAutocomplete-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-rs1at4-MuiFormControl-root-MuiTextField-root"
            data-testid="locationSearchBox"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1e3qgpi-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1nowbqt-MuiInputAdornment-root"
              >
                <span
                  aria-hidden="true"
                  class="notranslate"
                >
                  ​
                </span>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="SearchOutlinedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
                  />
                </svg>
              </div>
              <input
                aria-autocomplete="list"
                aria-expanded="false"
                aria-invalid="false"
                autocapitalize="none"
                autocomplete="off"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-qve9l1-MuiInputBase-input-MuiOutlinedInput-input"
                id="«r0»"
                placeholder="Budapest"
                role="combobox"
                spellcheck="false"
                type="text"
                value="Budapest, HU"
              />
              <div
                class="MuiAutocomplete-endAdornment css-1uhhrmm-MuiAutocomplete-endAdornment"
              >
                <button
                  aria-label="Open"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-2y136s-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                  tabindex="-1"
                  title="Open"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                    data-testid="ArrowDropDownIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-w4cd9x"
                >
                  <span
                    aria-hidden="true"
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
    </header>
    <div
      class="MuiBox-root css-g04nnx"
    >
      <span
        class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-wave css-jk665a-MuiSkeleton-root"
        data-testid="todaySkeleton"
        style="width: 350px; height: 375px;"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
